app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,274 @@
+
+            GregorianCalendar start = new GregorianCalendar();
+            GregorianCalendar end = new GregorianCalendar();
+            end.add(Calendar.YEAR, 100);
+            KeyPairGeneratorSpec keySpec = new KeyPairGeneratorSpec.Builder(context)
+                    // You'll use the alias later to retrieve the key.  It's a key for the key!
+                    .setAlias(SECRET_NAME)
+                    // The subject used for the self-signed certificate of the generated pair
+                    .setSubject(new X500Principal(String.format("CN=%s", SECRET_NAME)))
+                    // The serial number used for the self-signed certificate of the
+                    // generated pair.
+                    .setSerialNumber(BigInteger.valueOf(1337))
+                    // Date range of validity for the generated pair.
+                    .setStartDate(start.getTime())
+                    .setEndDate(end.getTime())
+                    .build();
+
+            generator.initialize(keySpec);
+            KeyPair pair = generator.generateKeyPair();
+
+            return trueForEncrypt ? pair.getPublic() : pair.getPrivate();
+        }
+    }
+
+    //Generate a random passphrase
+    public static byte[] generatePassphrase() {
+        Random random;
+        try {
+            //Use SecureRandom if possible (specifying algorithm for older versions of Android)
+            random = android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O ?
+                    SecureRandom.getInstanceStrong() : SecureRandom.getInstance("SHA1PRNG");
+        } catch (NoSuchAlgorithmException e) {
+            //Fallback to basic Random
+            random = new Random();
+        }
+
+        byte[] result = new byte[PASSPHRASE_LENGTH];
+
+        while (true) {
+            random.nextBytes(result);
+
+            //Make sure there are no zeroes in the passphrase
+            //SQLCipher passphrases must not contain any zero byte-values
+            //For more, see "Creating the Passphrase" section here:
+            //https://commonsware.com/Room/pages/chap-passphrase-001.html
+            boolean containsZero = false;
+            for (byte b : result) {
+                if (b == 0) {
+                    containsZero = true;
+                    break;
+                }
+            }
+
+            if (!containsZero) {
+                break;
+            }
+        }
+
+        return result;
+    }
+
+    private static String getTransformationString() {
+        String transformation;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+            transformation = String.format("%s/%s/%s", ALGORITHM, BLOCK_MODE, PADDING);
+        }
+        else {
+            transformation = "RSA/ECB/PKCS1Padding";
+        }
+
+        return transformation;
+    }
+
+    //Encryption backed by Android KeyStore
+    private static byte[] encrypt(Context context, byte[] bytes)
```

Comment:
Would be nice to write basic tests for `encrypt` and `decrypt` methods in this class.

Reply:
Done, I added a new EncryptionUtilsTest class with a test for the encrypt and decrypt functions. Note that the test generates its own RSA key pair to use, since the KeyStore isn't available in the test environment.